phpunit/experimental/interactivity-api/class-wp-directive-processor-test.php

```diff
@@ -0,0 +1,132 @@
+
+		$tags->next_tag( 'section' );
+		$this->assertSame( '<div><img>inside</div>', $tags->get_inner_html() );
+	}
+
+	public function test_set_inner_html_on_void_element_has_no_effect() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'img' );
+		$content = $tags->set_inner_html( 'This is the new img content' );
+		$this->assertFalse( $content );
+		$this->assertSame( self::HTML, $tags->get_updated_html() );
+	}
+
+	public function test_set_inner_html_sets_content_correctly() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'section' );
+		$tags->set_inner_html( 'This is the new section content.' );
+		$this->assertSame( '<div>outside</div><section>This is the new section content.</section>', $tags->get_updated_html() );
+	}
+
+	public function test_set_inner_html_updates_bookmarks_correctly() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'div' );
+		$tags->set_bookmark( 'start' );
+		$tags->next_tag( 'img' );
+		$this->assertSame( 'IMG', $tags->get_tag() );
+		$tags->set_bookmark( 'after' );
+		$tags->seek( 'start' );
+
+		$tags->set_inner_html( 'This is the new div content.' );
+		$this->assertSame( '<div>This is the new div content.</div><section><div><img>inside</div></section>', $tags->get_updated_html() );
+		$tags->seek( 'after' );
+		$this->assertSame( 'IMG', $tags->get_tag() );
+	}
+
+	public function test_set_inner_html_subsequent_updates_on_the_same_tag_work() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'section' );
+		$tags->set_inner_html( 'This is the new section content.' );
+		$tags->set_inner_html( 'This is the even newer section content.' );
+		$this->assertSame( '<div>outside</div><section>This is the even newer section content.</section>', $tags->get_updated_html() );
+	}
+
+	public function test_set_inner_html_followed_by_set_attribute_works() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'section' );
+		$tags->set_inner_html( 'This is the new section content.' );
+		$tags->set_attribute( 'id', 'thesection' );
+		$this->assertSame( '<div>outside</div><section id="thesection">This is the new section content.</section>', $tags->get_updated_html() );
+	}
+
+	public function test_set_inner_html_preceded_by_set_attribute_works() {
+		$tags = new WP_Directive_Processor( self::HTML );
+
+		$tags->next_tag( 'section' );
+		$tags->set_attribute( 'id', 'thesection' );
+		$tags->set_inner_html( 'This is the new section content.' );
+		$this->assertSame( '<div>outside</div><section id="thesection">This is the new section content.</section>', $tags->get_updated_html() );
+	}
+
+	/**
+	 * TODO: Review this, how that the code is in Gutenberg.
+	 */
+	public function test_set_inner_html_invalidates_bookmarks_that_point_to_replaced_content() {
+		$this->markTestSkipped( "This requires on bookmark invalidation, which is only in GB's WP 6.3 compat layer." );
```

Comment:
Oh, I almost forgot. @ockham, what should we do here?

Reply:
We would basically need to use the [WP 6.3 compat layer version of the Tag Processor](https://github.com/WordPress/gutenberg/blob/e740fff4ff73c5a10dd18afd9e0425a6742833a0/lib/compat/wordpress-6.3/html-api/class-gutenberg-html-tag-processor-6-3.php) for anything directive SSR related.

However, I'm not sure it's worth the hassle if we don't plan to get this into 6.3 anyway. In that case, we can just wait for 6.3 to be released and then remove this line.

---

I made a typo in the error message that's been bothering me for a while, and I forgot to fix it 😅 Not sure it's worth a separate PR though.

```suggestion
		$this->markTestSkipped( "This requires bookmark invalidation, which is only in GB's WP 6.3 compat layer." );
```

Reply:
> we can just wait for 6.3 to be released and then remove this line.

That sounds great 🙂

Thanks, Bernie!